Add blog post about new docker images (#195)
conan-io/conan-io.github.io: _posts/2021-08-09-Modern-docker-images.markdown

* Binaries can hardly be used outside of Conan Center because they need the newest version
of libstdc++ library that is not yet available in the official PPA. All executables built inside
ConanCenter won't work in the users' machines.
* A possible solution would be to statically link libstdc++ in all binaries, but this
solution has a number of risks.
* Each image uses the corresponding version of libstdc++ provided by the compiler.
* Better than the current scenario, where it is dependent on PPA and we have no control
over it.
* All images still use the same version of glibc, another advantage over the
current scenario.
* We will need to take care of the executables, as they will only be
compatible with later versions (as they are now).
* Better for users, than the current scenario.
The requirements related to libstdc++ are the same, but the glibc
version is the same version for all the packages.

Given the conditions and risks, we chose to go the second way: **Use the libstdc++ version
available together with the compiler**.

As Clang also supports ``libstdc++`` we choose ``libstdc++.so.6.0.28`` to be its default

> version. As commented before, that version has some advantages due its age and compatibility. To
> copy the library with Clang, first we needed to build GCC 10, then we could mount its Docker
> container and copy the ``libstdc++.so.6.0.28`` to the Clang image.

Reviewer: ```suggestion
version. As commented before, that version has some advantages due to its age and compatibility. It is also
the default for the latest LTS Ubuntu version, so we really think it should satisfy most of the use-cases.
```

Maybe remove those _implementation details_ related to docker images, and say a more _practical_ reason.

* From version to version, options are added or removed, reflecting the evolution of project
features and legacy deprecation. With these changes, it was inevitable to study the CMake files
of each version to understand which options do not work in subsequent versions or which option
should be used to specify the preferred library.
* Unlike GCC, LLVM has a huge range of parameters and a longer build time, around 1h depending
on the host. So, for each attempt, a long wait was needed to get the result.
* Until Clang 9 release, ``libc++`` was not automatically added to be linked when using Clang.
As a solution, the project supports a configuration file, where ``libc++`` can be specified by
default. However, this behavior changes between versions 6, 7 and 8, requiring different standards
and making it difficult to use the same Docker recipe for all versions.
* With the removal of the GCC dependency, it was necessary to use ``libunwind`` during the build.
It is already internalized in LLVM, but used as a dynamic library only. So a question arises, what
happens if a project uses the image with Clang and installs Conan's libunwind package? A big mess

> when linking, is the answer. Clang tries to link the version distributed by the Conan package,

Reviewer: ```suggestion
when linking is the answer. Clang tries to link the version distributed by the Conan package,
```

resulting in several errors. As a workaround, we renamed the original LLVM
``libunwind`` to ``libllvm-unwind``.

It became quite difficult to maintain from Clang 6 to 12 with the discovery of these issues and

> limitations.After a lot of discussions and advice from some of the LLVM maintainers we decided

Reviewer: ```suggestion
limitations. After a lot of discussions and advice from some of the LLVM maintainers we decided
```

Now, both the distribution version and the installed packages are configurable in terms of
version used. Previously, a script was used to update all 42 recipes as needed!

{% highlight docker %}

RUN printf '/usr/local/lib64\n' >> /etc/ld.so.conf.d/20local-lib.conf \
    && printf '/usr/local/lib\n' > /etc/ld.so .conf.d/20local-lib.conf \ ...

{% endhighlight %}

In order to not be affected by system packages or Conan packages that invoke ``apt-get``,
the compiler and its artifacts are installed in ``/usr/local``. However, this is not enough
to prioritize the order ``libstdc++`` used, for that we need to update ``ldconfig`` with the
local directories. Until then, this was not necessary in the previous images, as everything
was either consumed directly from the system, or installed directly in ``/usr``.

These are the main features of the base image, which is used in all final images.

For the construction of Clang, we tried to make it available from version 6.0 to 12, but we had a
series of obstacles and challenges that made us change our mind. Here we will share a little bit
of this long journey of CMake files and compilation hours. To see the full recipe, it is available

> [here](https://github.com/conan-io/conan-docker-tools/blob/master/modern/clang/Dockerfile).

Reviewer: Probably this is a leftover and doesn't belong to this section.

### Building GCC from Source

Now let's look at the GCC build image, the full recipe can be found
[here](https://github.com/conan-io/conan-docker-tools/blob/master/modern/gcc/Dockerfile),
but let's highlight a few points:

{% highlight docker %}

RUN cd gcc-${GCC_VERSION} \
    && ./configure --build=x86_64-linux-gnu --disable-bootstrap --disable-multilib ...

{% endhighlight %}

No matter the version, GCC continues to use the same lines for its build.  Some factors were
configured in this version used:

> * Bootstrap has been disabled to reduce build time to just 20 minutes.  * Fortran is enabled,
> but it barely increase the building time and final

Reviewer: New lines

```suggestion
* Bootstrap has been disabled to reduce build time to just 20 minutes.
* Fortran is enabled,
but it barely increase the building time and final
```

{% highlight docker %}

FROM ${DOCKER_USERNAME}/gcc${LIBSTDCPP_MAJOR_VERSION}-${DISTRO}:${DOCKER_TAG} as libstdcpp

FROM ${DOCKER_USERNAME}/base-${DISTRO}:${DOCKER_TAG} as deploy

ARG LIBSTDCPP_VERSION ARG LIBSTDCPP_PATCH_VERSION

ARG DOCKER_USERNAME ARG DOCKER_TAG ARG DISTRO

COPY --from=builder /tmp/install /tmp/clang COPY --from=libstdcpp /usr/local /tmp/gcc

RUN sudo mv /tmp/gcc/lib64 /usr/local/ \
    && sudo ln -s -f /usr/local/lib64/libstdc++.so.6.0.${LIBSTDCPP_PATCH_VERSION}
    /usr/local/lib64/libstdc++.so.6 \ && sudo ln -s -f /usr/local/lib64/libstdc++.so.6
    /usr/local/lib64/libstdc++.so \ && sudo cp -a /tmp/gcc/include/*
    /usr/local/include/ \ && sudo rm -rf /usr/lib/gcc/x86_64-linux-gnu/* \ && sudo cp -a
    /tmp/gcc/lib/gcc/x86_64-linux-gnu/${LIBSTDCPP_VERSION} /usr/lib/gcc/x86_64-linux-gnu/ \ &&
    sudo cp -a /tmp/gcc/lib/* /usr/local/lib/ \ ...

{% endhighlight %}

Similar to what was done with GCC, in Clang we also use the same base image and copy the
artifacts generated by the compiler to the ``/usr/local`` directory. However, the ``libstdc++``
library was extracted from the GCC 10 image. This is a necessity of the possible configurations

> supported by Conan.

Reviewer: ```suggestion
supported by Conan (possible values of ``compiler.libcxx`` setting).
```

> In addition Clang requires some interesting CMake definitions: * LLVM_ENABLE_PROJECTS: Only
> enable what we want, otherwise we will have tons of binaries and hours of build * LLVM_USE_LINKER:
> We enforce LLVM linker (lld). It's faster than GNU ld and reduces the total building time

Reviewer: New lines

```suggestion
In addition Clang requires some interesting CMake definitions: 
 * LLVM_ENABLE_PROJECTS: Only enable what we want, otherwise we will have tons of binaries and hours of build 
 * LLVM_USE_LINKER: We enforce LLVM linker (lld). It's faster than GNU ld and reduces the total building time
```

Since the beginning of the project, Conan Docker Tools has always used CI services such as Travis
and Azure. However, this did not give us the full power to prioritize the build in the queue,
customize the host, or customize the build lines to use Docker-in-Docker if necessary.

With that in mind we started using Jenkins to also build the new Docker images. The
big advantage in this is the use of features for cache in Docker. Previously a single
job took up to 2 hours if it was on other services without the use of caching. Now
using Jenkins and Docker ``--cache-from``, updating a package from base image to
final image takes just 4 minutes per job. The Jenkinsfile file used can be viewed
[here](https://github.com/conan-io/conan-docker-tools/blob/master/.ci/xenial.jenkinsfile).

Although the file looks complicated at first glance, it is still possible to use docker-compose
to build an image from scratch. As an example, let's use Clang 12:

{% highlight bash %}

$ cd modern $ docker-compose build base $ docker-compose build clang12-builder $ docker-compose
build clang12

{% endhighlight %}

The produced image will be named as ``conanio/clang12-ubuntu16.04:1.39.0``, where ``1.39.0``

> is the image tag and version of Conan installed. But it's totally configurable by the .env file.

Reviewer: ```suggestion
is the image tag and version of Conan installed. But it's totally configurable by the ``.env`` file.
```

   state. It was necessary to update the PPA address and build the images again.
 * The continuous integration service used, although it was considerably fast, it was not possible to
    customize and prioritize the build, if necessary.

Noticing the listed issues, we tried a radical solution.  This solution took more time to
implement but resulted in something better in terms of maintainability and practicality.

This being the case we decided to abandon PR #204 and start again from scratch, considering
the items listed above.

### A New Plan: Using the Same Base Image

We first need to understand the objective behind the project before we start implementing new
dockerfiles with their proper corrections. We want to understand what the expectation is for
these new images 2 years from now. With the current approach we are already maintaining over
40 different dockerfiles which can be problematic. We also want to address one of the  points
discussed previously: there will be rotation to avoid the accumulation of old images and
their restrictions in terms of maintenance.

#### The Plan

For current and already available images on hub.docker these will be kept but no longer supported

> (new versions will not be introduced). Your recipes and images will remain available, as their

Reviewer: ```suggestion
(new versions will not be introduced unless it's needed for ConanCenter). Your recipes and images will remain available, as their
```

release because it will be necessary to rebuild the packages that already exist in the Conan
Center and replace them to ensure full compatibility between packages. As for rotation and
maintenance, we believe it is necessary to rotate supported compilers over time, to avoid a
large build, effort and maintenance load for old images and packages that are not always used
by the community. Therefore, the following rule will be adopted:

* Clang will be supported from 10.0 to the newest version.
* GCC, on the other hand, is widely used for the Linux environment and we will generate images from version 5 to the latest.
* For both compilers we will keep updating all new compiler versions and Conan client version,
according to new releases.
* The multilib support was discarded as we are only interested in producing packages with
64-bit support.
* Fortran support has been added, thus producing ``gfortran`` together in
the image. Currently the Conan package for ``gfortran`` is totally broken and has a complex
dependencies chain to fix.

We always wanted to be independent of the Linux distribution so one of the problems we would like

> to solve is the compiler used and its libraries . In the initial pull request, we built the GCC

Reviewer: ```suggestion
to solve is the compiler used and its libraries. In the initial pull request, we built the GCC
```

{% highlight bash %}

$ cd modern && pytest tests --image conanio/gcc10-ubuntu16.04 --service deploy -vv --user 0:0

{% endhighlight %}

### The CI Service Change: From Travis to Azure and the Jenkins Arrival

Since the beginning of the project, Conan Docker Tools has always used CI services such as Travis
and Azure. However, this did not give us the full power to prioritize the build in the queue,
customize the host, or customize the build lines to use Docker-in-Docker if necessary.

With that in mind we started using Jenkins to also build the new Docker images. The
big advantage in this is the use of features for cache in Docker. Previously a single
job took up to 2 hours if it was on other services without the use of caching. Now
using Jenkins and Docker ``--cache-from``, updating a package from base image to
final image takes just 4 minutes per job. The Jenkinsfile file used can be viewed
[here](https://github.com/conan-io/conan-docker-tools/blob/master/.ci/xenial.jenkinsfile).

Although the file looks complicated at first glance, it is still possible to use docker-compose
to build an image from scratch. As an example, let's use Clang 12:

{% highlight bash %}

> $ cd modern $ docker-compose build base $ docker-compose build clang12-builder $ docker-compose
> build clang12

Reviewer: ```suggestion
$ cd modern 
$ docker-compose build base 
$ docker-compose build clang12-builder 
$ docker-compose build clang12
```

{% endhighlight %}

The produced image will be named as ``conanio/clang12-ubuntu16.04:1.39.0``, where ``1.39.0``
is the image tag and version of Conan installed. But it's totally configurable by the ``.env`` file.

In the case of legacy images, they will continue to be built in Azure when needed, we have no
intention of moving them to Jenkins due to effort and maintenance.

### How to build a Conan package with new Docker images

After building our new images, we are ready to build our Conan packages. Let's take Boost as
our example.

{% highlight shell %}

> $ docker run --rm -ti -v ${HOME}/.conan/data:/home/conan/.conan/data
> conanio/gcc10-ubuntu16.04:1.39.0 conan@148a77cfbc33:~$ conan install boost/1.76.0@ --build
> conan@148a77cfbc33:~$ exit

Reviewer: ```suggestion
$ docker run --rm -ti -v ${HOME}/.conan/data:/home/conan/.conan/data conanio/gcc10-ubuntu16.04:1.39.0
conan@148a77cfbc33:~$ conan install boost/1.76.0@ --build
conan@148a77cfbc33:~$ exit
```

{% endhighlight %}

Here, we start a temporary Docker container with interactive support. Also, we share our Conan
cache data as volume. After starting, we build Boost 1.76.0 and its dependencies from source. All
packages will be built and installed to the shared volume, so we can use it after closing the
container. To finish and remove the container, we just need to exit.

{% highlight shell %}

> $ docker run -d -t -v ${HOME}/.conan/data:/home/conan/.conan/data --name conan_container
> conanio/gcc10 $ docker exec conan_container conan install boost/1.76.0@ --build $ docker stop
> conan_container $ docker rm conan_container

Reviewer: ```suggestion
$ docker run -d -t -v ${HOME}/.conan/data:/home/conan/.conan/data --name conan_container conanio/gcc10 
$ docker exec conan_container conan install boost/1.76.0@ --build 
$ docker stop conan_container 
$ docker rm conan_container
```
